plugins/transformations/sequence2array/srcCxx/Sequence2ArrayTransformation.cxx

```diff
@@ -14,169 +14,147 @@
+#include "Sequence2ArrayTransformation.hpp"
 #include "DynamicDataHelpers.hpp"
 
 using namespace dds::core::xtypes;
 using dds::sub::SampleInfo;
 
 /*
- * @brief Check that both types are compatible, a compatible type for this
- * transformation will mean that all the elements are the same but the sequences
- * in the input are replaced by arrays in the output.
+ * @brief Check that both types are compatible. This function only works with
+ * unions and structs. A compatible type for this transformation will mean that
+ * all the elements are the same but the sequences in the input are replaced by
+ * arrays in the output.
  *
+ * This function has an indirect recursion. This function must be called only
+ * from are_types_compatible(). And also, this function call recursively to it.
  */
-bool Seq2ArrayTransformation::are_type_compatible(
+template<typename T>
+bool Sequence2ArrayTransformation::is_union_or_struct_type_compatible(
         const dds::core::xtypes::DynamicType & input_type,
         const dds::core::xtypes::DynamicType & output_type)
 {
-    bool is_compatible = true;
+    // Check that the types are the same
+    if (input_type.kind() != output_type.kind()) {
+        std::string error("the ouput field <" + output_type.name()
+                + "> is different to the input type <"
+                + input_type.name() + ">.");
+        throw std::runtime_error(error);
+    }
 
-    switch (input_type.kind().underlying()) {
-    case TypeKind::STRUCTURE_TYPE: {
-        if (output_type.kind() != TypeKind::STRUCTURE_TYPE) {
-            std::string error("Error: the ouput field <" + output_type.name()
-                    + "> is different to the input type <" + input_type.name());
-            throw std::runtime_error(error);
-        }
+    // check that the types are either union or struct
+    if (input_type.kind() != TypeKind::UNION_TYPE
+            && input_type.kind() != TypeKind::STRUCTURE_TYPE)  {
+        std::string error("unsupported kind of the input type <"
+            + input_type.name() + ">.");
+        throw std::runtime_error(error);
+    }
 
-        auto input_struct_type = static_cast<const StructType &>(input_type);
-        auto output_struct_type = static_cast<const StructType &>(output_type);
+    auto t_input_type = static_cast<const T &>(input_type);
+    auto t_output_type = static_cast<const T &>(output_type);
+    bool is_compatible = true;
 
-        if (input_struct_type.member_count() != output_struct_type.member_count()) {
-            std::string error("Error: different member count of input and output "
-                    "struct, input name: <" + input_struct_type.name()
-                    + ">, output name <" + output_struct_type.name() + ">: "
-                    + std::to_string(input_struct_type.member_count())
-                    + " != " + std::to_string(output_struct_type.member_count()));
+    // in case it is a union, check that the discriminator is the same
```

Comment:
```suggestion
    // in case it is a union, check that the discriminator is the same.
    // We rely on the overloaded `operator==` provided by DynamicType.
```

Reply:
done